File: app/src/main/java/com/example/administrator/essim/adapters/AuthorWorksAdapter.java
package com.example.administrator.essim.adapters;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.example.administrator.essim.R;
import com.example.administrator.essim.interf.OnItemClickListener;
import com.example.administrator.essim.response.IllustsBean;
import com.example.administrator.essim.utils.GlideUtil;

import java.util.List;

/**
 * Created by Administrator on 2018/3/23 0023.
 */
public class AuthorWorksAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private final int ITEM_TYPE_HEAD = 3;
    private final int ITEM_TYPE_BOTTOM = 2;
    private final int ITEM_TYPE_CONTENT = 1;
    private Context mContext;
    private LayoutInflater mLayoutInflater;
    private List<IllustsBean> mPixivRankItem;
    private OnItemClickListener mOnItemClickListener;

    public AuthorWorksAdapter(List<IllustsBean> item, Context context) {
        mPixivRankItem = item;
        mContext = context;
        mLayoutInflater = LayoutInflater.from(context);
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == ITEM_TYPE_HEAD) {
            return new ItemHolder(mLayoutInflater.inflate(R.layout.head_blank, parent, false));
        } else if (viewType == ITEM_TYPE_BOTTOM) {
            return new BottomViewHolder(mLayoutInflater.inflate(R.layout.bottom_refresh, parent, false));
        } else {
            return new PhotoHolder(mLayoutInflater.inflate(R.layout.pixiv_item_grid, parent, false));
        }
    }

    @Override
    public void onBindViewHolder(@NonNull final RecyclerView.ViewHolder holder, int p) {

        if (holder instanceof PhotoHolder) {
            int position = p - 1;
            Glide.with(mContext).load(new GlideUtil().getMediumImageUrl(mPixivRankItem.get(position)))
                    .into(((PhotoHolder) holder).mImageView);
            if (mPixivRankItem.get(position).isIs_bookmarked()) {
                ((PhotoHolder) holder).mImageView2.setImageResource(R.drawable.ic_favorite_white_24dp);
            } else {
                ((PhotoHolder) holder).mImageView2.setImageResource(R.drawable.ic_favorite_border_black_24dp);
            }
            if (mPixivRankItem.get(position).getPage_count() > 1) {
                ((PhotoHolder) holder).mTextView.setText(String.format("%sP", String.valueOf(mPixivRankItem.get(position).getPage_count())));
            } else {
                ((PhotoHolder) holder).mTextView.setVisibility(View.INVISIBLE);
            }
            if (mOnItemClickListener != null) {
                ((PhotoHolder) holder).itemView.setOnClickListener(v -> mOnItemClickListener.onItemClick(holder.itemView, position, 0));
                ((PhotoHolder) holder).mImageView2.setOnClickListener(v -> mOnItemClickListener.onItemClick(((PhotoHolder) holder).mImageView2, position, 1));
            }
        } else if (holder instanceof BottomViewHolder) {
            ((BottomViewHolder) holder).mTextView.setText("加载更多");
            ((BottomViewHolder) holder).mCardView.setOnClickListener(v -> mOnItemClickListener.onItemClick(holder.itemView, -1, 0));
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (position == 0) {
            return ITEM_TYPE_HEAD;
        } else if (position > 0 && position < mPixivRankItem.size() + 1) {
            return ITEM_TYPE_CONTENT;
        } else {
            return ITEM_TYPE_BOTTOM;
        }
    }

    private int getContentItemCount() {
        return mPixivRankItem.size();
    }

    public void setOnItemClickListener(OnItemClickListener mOnItemClickListener) {
        this.mOnItemClickListener = mOnItemClickListener;
    }

    @Override
    public int getItemCount() {
        if (mPixivRankItem.size() == 0) {
            return 0;
        } else {
            return getContentItemCount() + 2;
        }
    }

    private static class ItemHolder extends RecyclerView.ViewHolder {
        private ItemHolder(View itemView) {
            super(itemView);
        }
    }

    private static class PhotoHolder extends RecyclerView.ViewHolder {
        private ImageView mImageView, mImageView2;
        private TextView mTextView;

        private PhotoHolder(final View itemView) {
            super(itemView);
            mImageView = itemView.findViewById(R.id.pixiv_image);
            mImageView2 = itemView.findViewById(R.id.post_like);
            mTextView = itemView.findViewById(R.id.pixiv_item_size);
        }
    }

    private static class BottomViewHolder extends RecyclerView.ViewHolder {
        private CardView mCardView;
        private TextView mTextView;

        private BottomViewHolder(View itemView) {
            super(itemView);
            mTextView = itemView.findViewById(R.id.get_more_data);
            mCardView = itemView.findViewById(R.id.card_footer);
        }
    }
}

File: app/src/main/java/com/example/administrator/essim/adapters/AuthorWorksAdapter.kt
package com.example.administrator.essim.adapters

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewAnimationUtils
import android.view.ViewGroup
import com.bumptech.glide.Glide
import com.bumptech.glide.load.model.GlideUrl
import com.example.administrator.essim.R
import com.example.administrator.essim.interf.OnItemClickListener
import com.example.administrator.essim.response.IllustsBean
import com.example.administrator.essim.utils.GlideUtil
import kotlinx.android.synthetic.main.bottom_refresh.view.*
import kotlinx.android.synthetic.main.pixiv_item_grid.view.*

/**
 * Created by Administrator on 2018/3/23 0023.
 */

class AuthorWorksAdapter(private val mPixivRankItem: List<IllustsBean>,
                         private val mContext: Context) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val itemHead = 3
    private val itemBottom = 2
    private val itemContent = 1
    private var mOnItemClickListener: OnItemClickListener? = null
    private val mLayoutInflater: LayoutInflater = LayoutInflater.from(mContext)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder =
            when (viewType) {
                itemHead -> ItemHolder(mLayoutInflater.inflate(R.layout.head_blank, parent, false))
                itemBottom -> BottomViewHolder(mLayoutInflater.inflate(R.layout.bottom_refresh, parent, false))
                else -> PhotoHolder(mLayoutInflater.inflate(R.layout.pixiv_item_grid, parent, false))
            }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, p: Int) = when (holder) {
        is PhotoHolder -> {
            val position = p - 1
            Glide.with(mContext).load(GlideUtil().getMediumImageUrl(mPixivRankItem[position]))
                    .into(holder.itemView.pixiv_image)
            when {
                mPixivRankItem[position].isIs_bookmarked -> holder.itemView.post_like.setImageResource(R.drawable.ic_favorite_white_24dp)
                else -> holder.itemView.post_like.setImageResource(R.drawable.ic_favorite_border_black_24dp)
            }
            when {
                mPixivRankItem[position].page_count > 1 ->
                    holder.itemView.pixiv_item_size.text = String.format("%sP", mPixivRankItem[position].page_count.toString())
                else -> holder.itemView.pixiv_item_size.visibility = View.INVISIBLE
            }
            //单个item被点击，交给外界处理
            holder.itemView.setOnClickListener { mOnItemClickListener!!.onItemClick(holder.itemView, position, 0) }
            //设置收藏按钮的点击事件
            holder.itemView.post_like.setOnClickListener { v ->
                mOnItemClickListener!!.onItemClick(holder.itemView.post_like, position, 1)
                val anim = ViewAnimationUtils.createCircularReveal(holder.itemView,
                        v.x.toInt(),
                        v.y.toInt(),
                        0f, Math.hypot(holder.itemView.width.toDouble(),
                        holder.itemView.height.toDouble()).toFloat())
                anim.duration = 400
                anim.start()
            }
            //设置收藏按钮的长按事件
            holder.itemView.post_like.setOnLongClickListener { v ->
                mOnItemClickListener!!.onItemLongClick(holder.itemView.post_like, position)
                val anim = ViewAnimationUtils.createCircularReveal(holder.itemView,
                        v.x.toInt(),
                        v.y.toInt(),
                        0f, Math.hypot(holder.itemView.width.toDouble(),
                        holder.itemView.height.toDouble()).toFloat())
                anim.duration = 400
                anim.start()
                true
            }
        }
        is BottomViewHolder -> {
            holder.itemView.get_more_data.text = "加载更多"
            holder.itemView.card_footer.setOnClickListener { mOnItemClickListener!!.onItemClick(holder.itemView, -1, 0) }
        }
        else -> {
        }
    }

    override fun getItemViewType(position: Int): Int = when (position) {
        0 -> itemHead
        else -> if (position > 0 && position < mPixivRankItem.size + 1) {
            itemContent
        } else {
            itemBottom
        }
    }

    fun setOnItemClickListener(onItemClickListener: OnItemClickListener) {
        mOnItemClickListener = onItemClickListener
    }

    override fun getItemCount(): Int = when {
        mPixivRankItem.isEmpty() -> 0
        else -> mPixivRankItem.size + 2
    }

    private class ItemHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
    private class PhotoHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
    private class BottomViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)
}
